src/hathor.c

```diff
@@ -92,64 +68,111 @@ void compress_public_key(unsigned char *value) {
     value[0] = ((value[64] & 1) ? 0x03 : 0x02);
 }
 
-int encode_base58(const unsigned char *in, size_t inlen, unsigned char *out, size_t outlen) {
-    unsigned char buffer[B58_MAX_INPUT_SIZE * 138 / 100 + 1] = {0};
-    size_t i = 0, j;
-    size_t startAt, stopAt;
-    size_t zeroCount = 0;
-    size_t outputSize;
+void pubkey_hash_to_address(uint8_t *hash, uint8_t *out) {
+    unsigned char checksum_buffer[32];
+    // prepend version
+    out[0] = P2PKH_VERSION_BYTE;
+    os_memmove(out+1, hash, 20);
+    // sha256d of above and get first 4 bytes (checksum)
+    sha256d(out, 21, checksum_buffer);
+    os_memmove(out+21, checksum_buffer, 4);
+}
 
-    if (inlen > B58_MAX_INPUT_SIZE) {
-        return -1;
-    }
+void pubkey_to_address(cx_ecfp_public_key_t *public_key, uint8_t *out) {
+    unsigned char hash_buffer[20];
+    // get compressed pubkey
+    compress_public_key(public_key->W);
+    // get hash160
+    hash160(public_key->W, 33, hash_buffer);
+    pubkey_hash_to_address(hash_buffer, out);
+}
 
-    while ((zeroCount < inlen) && (in[zeroCount] == 0)) {
-        ++zeroCount;
+/**
+ * Validates that a script has the format of P2PKH. Throws an exception if doesn't.
+ * P2PKH scripts have the format:
+ *   [OP_DUP, OP_HASH160, pubkey_hash_len, pubkey_hash, OP_EQUALVERIFY, OP_CHECKSIG]
+ * Considering that pubkey hashes have 20 bytes and the values of the opcodes:
+ *   [0x76, 0xA9, 20, pubkey_hash, 0x88, 0xAC]
+ */
+void validate_p2pkh_script(uint8_t *in) {
```

Comment:
This algorithm partially works because it is not considering the timelock.

That's the correct regexp we should match: `^(?:(DATA_4) OP_GREATERTHAN_TIMESTAMP)? OP_DUP OP_HASH160 (DATA_20) OP_EQUALVERIFY OP_CHECKSIG$`.

Reply:
Added a comment explaining that we don't support timelock yet.